src/zarr/core/common.py

```diff
@@ -67,6 +68,132 @@ class NamedConfig(TypedDict, Generic[TName, TConfig]):
     """The configuration of the object."""
 
 
+class NamedConfig(TypedDict, Generic[TName, TConfig]):
+    """
+    A typed dictionary representing an object with a name and configuration, where the configuration
+    is a mapping of string keys to values, e.g. another typed dictionary or a JSON object.
+
+    The configuration key is not required.
+
+    This class is generic with two type parameters: the type of the name (``TName``) and the type of
+    the configuration (``TConfig``).
+    """
+
+    name: ReadOnly[TName]
+    """The name of the object."""
+
+    configuration: ReadOnly[NotRequired[TConfig]]
+    """The configuration of the object."""
+
+
+class ArrayMetadataJSON_V2(TypedDict):
+    """
+    A typed dictionary model for Zarr V2 array metadata.
+    """
+
+    zarr_format: Literal[2]
+    dtype: str | StructuredName_V2
+    shape: Sequence[int]
+    chunks: Sequence[int]
+    dimension_separator: NotRequired[Literal[".", "/"]]
+    fill_value: Any
+    filters: Sequence[CodecJSON_V2[str]] | None
+    order: Literal["C", "F"]
+    compressor: CodecJSON_V2[str] | None
+    attributes: NotRequired[Mapping[str, JSON]]
+
+
+class GroupMetadataJSON_V2(TypedDict):
+    """
+    A typed dictionary model for Zarr V2 group metadata.
+    """
+
+    zarr_format: Literal[2]
+    attributes: NotRequired[Mapping[str, JSON]]
+    consolidated_metadata: NotRequired[ConsolidatedMetadata_JSON_V2]
+
+
+class ArrayMetadataJSON_V3(TypedDict):
+    """
+    A typed dictionary model for Zarr V3 array metadata.
+    """
+
+    zarr_format: Literal[3]
+    node_type: Literal["array"]
+    data_type: str | NamedConfig[str, Mapping[str, object]]
+    shape: Sequence[int]
+    chunk_grid: NamedConfig[str, Mapping[str, object]]
+    chunk_key_encoding: NamedConfig[str, Mapping[str, object]]
+    fill_value: object
+    codecs: Sequence[str | NamedConfig[str, Mapping[str, object]]]
+    attributes: NotRequired[Mapping[str, object]]
+    storage_transformers: NotRequired[Sequence[NamedConfig[str, Mapping[str, object]]]]
+    dimension_names: NotRequired[Sequence[str | None]]
+
+
+class GroupMetadataJSON_V3(TypedDict):
+    """
+    A typed dictionary model for Zarr V3 group metadata.
+    """
+
+    zarr_format: Literal[3]
+    node_type: Literal["group"]
+    attributes: NotRequired[Mapping[str, JSON]]
+    consolidated_metadata: NotRequired[ConsolidatedMetadata_JSON_V3 | None]
+
+
+# TODO: use just 1 generic class and parametrize the type of the value type of the metadata
+# I.e., ConsolidatedMetadata_JSON[ArrayMetadataJSON_V2 | GroupMetadataJSON_V2]
+class ConsolidatedMetadata_JSON_V2(TypedDict):
+    """
+    A typed dictionary model for Zarr consolidated metadata.
+
+    This model is parameterized by the type of the metadata itself.
+    """
+
+    kind: Literal["inline"]
+    must_understand: Literal["false"]
```

Comment:
Is this correct for v2 consolidated metadata? IIRC, it didn't have `must_understand` and maybe had a `version` field? I guess maybe I'm mixing up consolidated metadata as written by zarr-python 2.x, and the v3 spec.